### PR TITLE
feat: emails in a celery task, for calls to `EnterpriseCustomer#notify_enrolled_learners()`

### PR DESCRIPTION
**Pull request by binodpant**
## Included in PR
Summary of changes (updated from what I had since last round of review):
- Added a task to send email notifications in `enterprise/tasks.py`
- Extracted out a method to only serialize the email_items list so that we can use it to call a celery task
- Now calling the celery task from the model's `notify_enrolled_learners` function so every caller of the notify_enrolled_learners function gets this ability, not just bulk enrollment
- Moved a function `parse_lms_api_datetime` to utils. Strictly speaking not needed for this PR but I decided to leave it there 

For bulk enrollment, no. of tasks created = no. of courses. Each task will email all learners for enrollment into that course

ENT-4779

## Testing notes:

* ADDED: Unit test for the new task
* UPDATED: Update tests that call the view to ensure the task is invoked correctly

`Staging`:
Once this (via edx-platform) gets to staging, test plan is to put a pause on edx-platform prod pipeline, run a test with enrollment of these kinds:

- bulk enroll with emails on and off
- regular enroll in a course from courseware page
- learner portal enroll in a course

`Manual`:

Executed bulk enrollment, and verified a task was fired for each course and also an email file was generated per course 

```
ls -al /edx/src/ace_messages/
total 1296
drwxr-xr-x 5 root root    160 Jul 28 03:16 .
drwxr-xr-x 4 root root    128 Jan 14  2021 ..
-rw-r--r-- 1 root root 438845 Jul 28 03:16 20210728-031634-139685836518400.log
-rw-r--r-- 1 root root 438843 Jul 28 03:16 20210728-031636-139685832905728.log
-rw-r--r-- 1 root root 440297 Jul 28 03:16 20210728-031638-139685848429376.log
```

in logs I see 

```
2021-07-29 13:38:53,667 INFO 32082 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:315 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: edx>, <User: verified>} into course course-v1:edX+909878+2T2021
2021-07-29 13:38:54,398 INFO 32082 [celery.app.trace] [user 17] [ip 172.18.0.16] trace.py:125 - Task enterprise.tasks.send_enterprise_email_notification[0c160b49-e277-4654-9a8c-da6a3bd78bf8] succeeded in 0.042356400001153816s: None
2021-07-29 13:38:54,399 INFO 32082 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:315 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: edx>, <User: verified>} into course course-v1:edX+303707808+2T2021
2021-07-29 13:38:55,056 INFO 32082 [celery.app.trace] [user 17] [ip 172.18.0.16] trace.py:125 - Task enterprise.tasks.send_enterprise_email_notification[a0583741-2b13-4e00-8a66-5685bb492fd5] succeeded in 0.043135300002177246s: None
2021-07-29 13:38:55,057 INFO 32082 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:315 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: edx>, <User: verified>} into course course-v1:edX+808404707+2T2021
2021-07-29 13:38:55,740 INFO 32082 [celery.app.trace] [user 17] [ip 172.18.0.16] trace.py:125 - Task enterprise.tasks.send_enterprise_email_notification[4a8eb86b-e88d-4289-a6a0-28ce30d3468a] succeeded in 0.04192120000516297s: None
[29/Jul/2021 13:38:55] "POST /enterprise/api/v1/enterprise-customer/9a5afb8e-63bc-456a-a600-81b60777e5aa/enroll_learners_in_courses/ HTTP/1.1" 201 789
```

Each file represents emails send to all 25 learners I used in testing.

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/api/v1/views.py [hunk 1]
```diff
@@ -318,12 +319,13 @@ def enroll_learners_in_courses(self, request, pk):
                 ))
                 track_enrollment(PATHWAY_CUSTOMER_ADMIN_ENROLLMENT, request.user.id, course_run)
                 if serializer.validated_data.get('notify'):
-                    enterprise_customer.notify_enrolled_learners(
-                        catalog_api_user=request.user,
-                        course_id=course_run,
-                        users=pending_users | existing_users,
-                        admin_enrollment=True,
+                    email_items = enterprise_customer.prepare_notification_content(
```
Comment (edx-abolger) [inline enterprise/api/v1/views.py:322]: Nit: can you change the name of 'prepare_notification_content' to 'serialize_notification_content' ?

File: enterprise/models.py [hunk 2]
```diff
@@ -723,20 +733,51 @@ def notify_enrolled_learners(self, catalog_api_user, course_id, users, admin_enr
...
+                    'type': 'course',
+                    'start': course_start,
+                },
+                "dashboard_url": dashboard_url,
+                "enterprise_customer_uuid": self.uuid,
+                "admin_enrollment": admin_enrollment,
+            })
+        return email_items
+
+    def notify_enrolled_learners(self, catalog_api_user, course_id, users, admin_enrollment=False):
```
Comment (edx-abolger) [inline enterprise/models.py:754]: Don't name the notify_enrolled_learners task and this function the same thing, add _task if you are going to keep both items in 2 separate modules.

File: enterprise/utils.py [hunk 3]
```diff
@@ -455,20 +455,40 @@ def find_enroll_email_template(enterprise_customer, template_type):
     return enterprise_template_config
 
 
+def create_dict_from_user(user):
```
Comment (edx-abolger) [inline enterprise/utils.py:458]: Are you sure that the Django user object doesn't already have a serializer? Home writing one like this adds complexity. Can you look at
```

from django.forms.models import model_to_dict
model_to_dict(instance)
```
function and see if it meets your needs here?

**Review by edx-abolger — CHANGES_REQUESTED** [review]
I understand caution, but you should go ahead and make the new task used everywhere, not just for bulk enrollment. The conservative approach here actually makes it harder to understand, I would rather place that conservative approach into adding a unit test for your new task (which you should make one of anyway). The module circular dependency can be solved a different way, let me know if you need help on that.

File: enterprise/utils.py (hunk 3 above)
Comment (binodpant) [reply enterprise/utils.py:458]: thanks, thinks I should be able to make that work and avoid a custom function, will be back!

**Comment by binodpant** [comment]
> I understand caution, but you should go ahead and make the new task used everywhere, not just for bulk enrollment. The conservative approach here actually makes it harder to understand, I would rather place that conservative approach into adding a unit test for your new task (which you should make one of anyway). The module circular dependency can be solved a different way, let me know if you need help on that.

Sure thing! I was, like you said, being risk averse. There is opportunity to apply the tasking to other emails sent out too and get it done in this round. I do see the value of doing it now rather than later, so all the emails get this benefit, it isn't that big a stretch to apply and test it out by enrolling directly from the courseware, learner portal etc in addition to enrolling from bulk enroll page. Yes the task portion also needs some unit testing, so going to look into that. So I accept your proposal. Be back with this!

**Comment by binodpant** [comment]
also on the circular dep looks like I might already have worked myself out of that, but if I get stuck or sth will let you know thanks

File: enterprise/utils.py [hunk 4]
```diff
@@ -1962,3 +2069,28 @@ def get_best_mode_from_course_key(course_key):
         else CourseModes.AUDIT
 
     return best_course_mode
+
+
+def parse_lms_api_datetime(datetime_string, datetime_format=LMS_API_DATETIME_FORMAT):
```
Comment (binodpant) [inline enterprise/utils.py:2074]: I moved it here to utils since calling EnterpriseCatalogApiClient  from tasks was causing a circular dep. I think this is a better location for this function anyway

**Comment by binodpant** [comment]
A final round of testing manually as well as fixing up one test failure remains but I largely made the changes I needed to make I think, so that now I can keep the model function in place and also this change to use celery will apply to all calls of that function

File: tests/test_enterprise/api/test_views.py [hunk 5]
```diff
@@ -3339,7 +3339,8 @@ class TestBulkEnrollment(BaseTestEnterpriseAPIViews):
                     'email': 'abc@test.com',
                     'course_run_key': 'course-v1:edX+DemoX+Demo_Course',
                     'license_uuid': '5a88bdcade7c4ecb838f8111b68e18ac'
-                }]
+                }],
+                'notify': 'true',
```
Comment (binodpant) [inline tests/test_enterprise/api/test_views.py:3343]: inserting this here so we can hit the code path that sends notifications

File: enterprise/tasks.py [hunk 6]
```diff
@@ -8,25 +8,60 @@
...
+                item['dashboard_url'],
+                enterprise_customer_uuid,
+                email_connection=email_conn,
+                admin_enrollment=admin_enrollment,
+            )
+
+
 @shared_task
 @set_code_owner_attribute
 def create_enterprise_enrollment(course_id, enterprise_customer_user_id):
```
Comment (edx-abolger) [inline enterprise/tasks.py:55]: Why were these changes needed? This is an unrelated shared task to the one you are modifying?

Comment (binodpant) [reply enterprise/tasks.py:55]: oh these were because the direct use of the models in this task was causing circular ref errors since the model now calls the task

Comment (binodpant) [reply enterprise/tasks.py:55]: ah just read your latest comment, yes will do that thanks

File: enterprise/utils.py [hunk 7]
```diff
@@ -455,20 +460,121 @@ def find_enroll_email_template(enterprise_customer, template_type):
...
+            site=lms_root_url,
+            login_or_register='{login_or_register}',  # We don't know the value at this time
+            course_path=course_path
+        )
+
+    try:
+        course_start = parse_lms_api_datetime(course_details.get('start'))
+    except (TypeError, ValueError):
+        course_start = None
+        LOGGER.exception(
```
Comment (edx-abolger) [inline enterprise/utils.py:537]: +1 for using Logger.exception so we get the stack trace!!!

File: tests/test_enterprise/api/test_views.py [hunk 8]
```diff
@@ -3469,6 +3473,38 @@ def test_bulk_enrollment_in_bulk_courses_pending_licenses(
         else:
             mock_track_enroll.assert_not_called()
 
+        # verify notification sent correctly for each course to applicable learners
+        if 'notify' in body:
```
Comment (edx-abolger) [inline tests/test_enterprise/api/test_views.py:3477]: nit: This is already a long test and I don't love conditional logic in tests, it makes them breakable. Can you please refactor the existing test by splitting it in half (old_test = 1 new function, if conditional code into 2nd new function) and then make it like this: 

```

def _original_test_logic():
   stuff

def notify_test():
   your new code

@<data>
def test_bulk_enrollment_in_bulk_courses_pending_licenses():
  original_code(params)
  
@<data you want to use for notifications>
def test_bulk_enrollment_in_bulk_courses_pending_licenses_with_notifications():
  original_code(params)
  new_notification_code(params)

```


Another option is to just run the notify verification on every test case.

File: tests/test_enterprise/test_tasks.py [hunk 9]
```diff
@@ -99,3 +105,38 @@ def test_create_enrollment_task_no_create_duplicates(self, catalog_api_client_mo
...
+        admin_enrollment = True
+
+        mail_conn = mock.MagicMock()
+        mock_email_conn.return_value.__enter__.return_value = mail_conn
+
+        email_items = serialize_notification_content(
+            enterprise_customer,
+            course_details,
+            self.FAKE_COURSE_ID,
+            users + [pending_user],
```
Comment (edx-abolger) [inline tests/test_enterprise/test_tasks.py:126]: I like that both cases are covered here!

**Review by edx-abolger — CHANGES_REQUESTED** [review]
2 small remaining requests, almost ready to go, looks very clean.

File: tests/test_enterprise/api/test_views.py (hunk 8 above)
Comment (binodpant) [reply tests/test_enterprise/api/test_views.py:3477]: I like this! let me get back with this change, so the notification use case testing is separate

Comment (binodpant) [reply tests/test_enterprise/api/test_views.py]: done!

**Review by edx-abolger — APPROVED** [review]
LGTM!

File: enterprise/tasks.py [hunk 10]
```diff
@@ -8,25 +8,60 @@
...
+        *
+        * email_items: list of dictionary objects with keys:
+        *   user (dict): a dict with either of the following forms:
+              - 1: { 'first_name': name, 'username': user_name, 'email': email } (similar to a User object)
+              - 2: { 'user_email' : user_email } (similar to a PendingEnterpriseCustomerUser object)
+        *   enrolled_in (dict): name and optionally other keys needed by templates
+        *   dashboard_url (str)
+    """
+    with mail.get_connection() as email_conn:
+        for item in email_items:
```
Comment (macdiesel) [inline enterprise/tasks.py:42]: Since this isn't wrapped with a try/catch, I believe that an error in one email send will prevent all from being sent. Is this behavior OK for your use case?

**Review by macdiesel — COMMENTED** [review]
One other comment about this. Do you need to be concerned about messages that fail? I'm perhaps you might want to [add a retry](https://docs.celeryproject.org/en/stable/userguide/tasks.html#retrying) in case the error is transient? 

Would you want to retry sending these emails later should they failure? If so I would suggest saving these messages off to a dead letter queue so they could be re-run once the code/infra/whatever is fixed at a later point in time. 

Perhaps beyond the MVP for this but right now with the synchronous nature of this call the user will know when the send fails via the webapp. Since this is being moved to an async task the user will no longer have this feedback. This could cause customer confusion and potentially increase tickets to ECS.

File: enterprise/tasks.py (hunk 10 above)
Comment (edx-abolger) [reply enterprise/tasks.py:42]: probably not ok. May as well fix that here, with at least a logged exception, it existed in the prior version but that doesn't mean it wasn't a problem.  For the beyond MVP stuff we want to move this to Braze, so I wouldn't try to provide failed send feedback here.

Comment (binodpant) [reply enterprise/tasks.py:42]: Sorry I had merged this by the time I read this. But yeah agree with Brian's observation. I can create a separate PR to do that. And I will likely merge and test that version tomorrow, which should make this sturdier

Comment (binodpant) [reply enterprise/tasks.py:42]: I don't know exactly what Exceptions are going to be thrown by the underlying code yet though.. a quick look at the mail.send() api reveals nothing. will a `except Exception be` the safest option?

Comment (macdiesel) [reply enterprise/tasks.py:42]: That hammer will catch it all, sounds good to me.

Comment (binodpant) [reply enterprise/tasks.py:42]: to close the loop this update is now being released via https://github.com/edx/edx-platform/pull/28324
